universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java

```diff
@@ -189,23 +189,38 @@ public ProgramDefinition moveBlock(long blockId, Direction direction)
         blockSlice.startsBefore(otherBlockSlice) ? blockSlice : otherBlockSlice;
     BlockSlice latterSlice =
         blockSlice.startsBefore(otherBlockSlice) ? otherBlockSlice : blockSlice;
+
+    // An illegal move is when a predicate in a later block depends on an earlier block.
+    // Get the questions used in any predicate in the latter block.
+    ImmutableSet<Long> predicateQuestions =
+        blockDefinitions().subList(latterSlice.startIndex(), latterSlice.endIndex()).stream()
+            .filter(b -> b.visibilityPredicate().isPresent())
+            .flatMap(b -> b.visibilityPredicate().get().getQuestions().stream())
+            .collect(toImmutableSet());
+
+    if (!predicateQuestions.isEmpty()) {
+      if (blockDefinitions().subList(earlierSlice.startIndex(), earlierSlice.endIndex()).stream()
+          .flatMap(b -> b.programQuestionDefinitions().stream().map(ProgramQuestionDefinition::id))
+          .anyMatch(predicateQuestions::contains)) {
```

Comment:
(super nit) this would call `Set#contains` for each id. Instead you could turn these ids into a set and check for an intersect. This is def premature optimization.

Reply:
Implemented the `hasValidPredicateOrdering` method, which renders this code obsolete